src/connector/codec/src/decoder/json/mod.rs

```diff
@@ -12,11 +12,170 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use std::collections::HashMap;
+use std::{fs, mem};
+
 use anyhow::Context;
 use risingwave_pb::plan_common::ColumnDesc;
+use serde_json::Value;
+use thiserror::Error;
+use url::Url;
 
 use super::avro::{avro_schema_to_column_descs, MapHandling};
 
+#[derive(Debug, Error, thiserror_ext::ContextInto)]
+pub enum Error {
+    #[error("could not open schema from {filename}")]
+    SchemaFromFile {
+        filename: String,
+        source: std::io::Error,
+    },
+    #[error("parse error for url {url}")]
+    UrlParse {
+        url: String,
+        source: url::ParseError,
+    },
+    #[error("schema from {url} not valid JSON")]
+    SchemaNotJson { url: String, source: std::io::Error },
+    #[error("request error")]
+    Request { url: String, source: reqwest::Error },
+    #[error("schema from {url} not valid JSON")]
+    SchemaNotJsonSerde {
+        url: String,
+        source: serde_json::Error,
+    },
+    #[error("ref `{ref_string}` can not be resolved as pointer, `{ref_fragment}` can not be found in the schema")]
+    JsonRefPointerNotFound {
+        ref_string: String,
+        ref_fragment: String,
+    },
+    #[error("json ref error")]
+    JSONRef {
+        #[from]
+        source: std::io::Error,
+    },
+    #[error("need url to be a file or a http based, got {url}")]
+    UnsupportedUrl { url: String },
+    #[error(transparent)]
+    Uncategorized(
+        #[from]
+        #[backtrace]
+        anyhow::Error,
+    ),
+}
+
+type Result<T, E = Error> = std::result::Result<T, E>;
+
+#[derive(Debug)]
+pub struct JsonRef {
```

Comment:
Acknowledge that `$id` and `$anchor` are currently ignored and unsupported.

Reply:
I did check the jsonschema library. It's more like designed for validation and is vary hard to use it for dereffing the schema file.

Reply:
shall we move the impl to a standalone crate? I feel that the kernel should not touch the logic of JsonRef.

Reply:
No. We may need to hack it to support confluent schema registry. The confluent schema registry should be handled differently. IIUC they can use reference name to retrieve a reference instead of URI.